### PR TITLE
Replaced many `exit(1)`s with exception raising

### PR DESCRIPTION
**Pull request by inclement**
This both makes the code neater and will make it easier to test things.

### CODE REVIEW

**Comment by AndreMiras** [comment]
I was wondering if `test_invalid_recipe_order_and_bootstrap` would pass, but indeed it didn't :confused:

File: pythonforandroid/toolchain.py [hunk 1]
```diff
@@ -58,8 +59,7 @@ def check_python_dependencies():
                 ok = False
 
     if not ok:
-        print('python-for-android is exiting due to the errors above.')
-        exit(1)
+        print('python-for-android is exiting due to the errors logged above')
```
Comment (JonasT) [inline pythonforandroid/toolchain.py]: Should there be an error raised here? Or is this change intended?

Comment (inclement) [reply pythonforandroid/toolchain.py]: I originally did, but actually this code is run before most of the build process (because it has to run before the imports) so I think it needs a different treatment.

**Comment by inclement** [comment]
@AndreMiras Just rebased this to fix merge conflicts, any objections to merging it (if the tests pass)?

**Comment by JonasT** [comment]
Just to barge in, I have been eyeing this for a while and I think it's a really good code smell fix. Haven't commented so far since I haven't had the chance to test if it actually doesn't break stuff, but it looks like a thing really worth improving

File: pythonforandroid/toolchain.py (hunk 1 above)
Comment (JonasT) [reply pythonforandroid/toolchain.py]: well but currently, this seems to be a big change in behavior since the code flow just seems to resume. Is that intended? Maybe the `exit` should be left in if a runtime error doesn't make sense here? Or will this problem of `not ok` be evaluated again & properly dealt with later?

Comment (inclement) [reply pythonforandroid/toolchain.py]: I also added the exit(1) back in with the previous comment, I missed that I'd accidentally removed it before so thanks for pointing it out!

Comment (JonasT) [reply pythonforandroid/toolchain.py]: Ah right, the removing line is gone now. Didn't see that :laughing:

**Comment by AndreMiras** [comment]
Hi @inclement can you give me write access to your `improve_fault_handling` branch so I can help to finish with this one?
We're almost there, I would be a pity to wait for that PR to conflict with other changes

File: pythonforandroid/distribution.py [hunk 2]
```diff
@@ -129,6 +128,7 @@ def get_distribution(cls, ctx, name=None, recipes=[],
         # If there was a name match but we didn't already choose it,
         # then the existing dist is incompatible with the requested
         # configuration and the build cannot continue
+<<<<<<< HEAD
```
Comment (AndreMiras) [inline pythonforandroid/distribution.py]: Looks like something went wrong during the rebasing :confused: 
Apparently I have write access to the branch so I'll pick this up

Comment (AndreMiras) [reply pythonforandroid/distribution.py]: Fixed
